feat(storage): thread new storage reference through all read requests

GeorgeMac opened the pull request:
Fixes: FLI-724

This is a combination refactor and addition of a new `reference` field on all read requests.

In order to thread this everywhere with an empty default, I packaged up many of the storage method call arguments into structures. In particular, I created a few common structures.

`NamespaceRequest` is used for operations which only need a namespace:
https://github.com/flipt-io/flipt/blob/a9368638a19eef721db07b9929627225101372ca/internal/storage/storage.go#L373-L378

`ResourceRequest` is used for any operation which is scoped to some resource (e.g. scoped by flag or segment):
https://github.com/flipt-io/flipt/blob/a9368638a19eef721db07b9929627225101372ca/internal/storage/storage.go#L408-L413

Each of these types embed a reference which can be empty (zero state).
In a subsequent PR I will start to depend upon these references.
However, the server layer has been updated in this change to start threading these references from the RPC layer into the storage layer.

markphelps reviewed (approved):
amazing refactor!!

![](https://media.giphy.com/media/3o7aDeFfrdxWOYnx72/giphy.gif)

GeorgeMac commented:
@markphelps I just renamed a bunch of these new things and added some additional doc comments.
I mostly renamed things from "predicate" to "request" as I felt it was too unecessarilly fancy talk 🎩 .

GeorgeMac commented:
Also, this last commit allows the existing fs stores to skip taking a reference and ignoring it for now.
Instead, there is an adaptor that does the dropping reference part and log if its non-empty.
The Git and OCI stores will ultimately implement the `ReferencedSnapshotStore` interface and support multiple references that way. But the others will warn if you try to access no default references through the adaptor.

markphelps reviewed (dismissed):
wicked

GeorgeMac commented:
Last pass @markphelps 

I rebased and fixed this up after the the GCS changes and the cache changes 👍

yquansah commented on internal/storage/storage.go:
```diff
@@ -299,13 +301,150 @@ func ListWithQueryParamOptions[T any](opts ...QueryOption) ListOption[T] {
+	p := IDRequest{ID: id}
+	containers.ApplyAll(&p.ReferenceRequest, opts...)
+	return p
+}
```
Nice abstractions!

yquansah reviewed (approved):
This looks good to me! Nice work.
